src/puppetlabs/puppetdb/http/command.clj

```diff
@@ -168,32 +167,38 @@
               (normalize-new-request req)
               (normalize-old-request req)))))
 
-(defn- realize-body [body max-command-size]
+(defn- stream-with-max-check
   "Returns the body as an in-memory string or byte-array, reading it
   if necessary.  Throws ::body-stream-overflow if the max-command-size
   is not false and not respected."
+  [body max-command-size]
   (if-not max-command-size
-    (if (instance? java.io.InputStream body)
-      (IOUtils/toByteArray body)
-      body)
     (cond
       (instance? java.io.InputStream body)
-      (IOUtils/toByteArray
-       (restrained-drained-stream body (long max-command-size)))
+      body
+
+      (string? body)
+      (java.io.ByteArrayInputStream. (.getBytes body "UTF-8"))
+
+      :else
+      (throw (Exception. (str "Unexpected body type: " (class body)))))
+    (cond
+      (instance? java.io.InputStream body)
+      (restrained-drained-stream body (long max-command-size))
```

Comment:
It seems like we should revisit this behavior at some point. At the time, it was a memory issue. Now it's not an issue of enqueuing, but rather an issue of dequeuing. We could stream it to disk, then move it to the DLO if it was too large. This also opens up a potential DoS, what if people throw lots of large messages our way and fill up the disk. Something to think about here.